src/main/java/com/cloudbees/jenkins/plugins/sshagent/SSHAgentStepExecution.java

```diff
@@ -0,0 +1,244 @@
+            } catch (Throwable th) {
+                context.onFailure(th);
+            }
+            execution.purgeSockets();
+            context.onFailure(t);
+        }
+
+    }
+
+    private static final class ExpanderImpl extends EnvironmentExpander {
+
+        private static final long serialVersionUID = 1L;
+
+        private final SSHAgentStepExecution execution;
+
+        ExpanderImpl(SSHAgentStepExecution execution) {
+            this.execution = execution;
+        }
+
+        @Override
+        public void expand(EnvVars env) throws IOException, InterruptedException {
+            env.override("SSH_AUTH_SOCK", execution.getSocket());
+        }
+    }
+
+    /**
+     * Initializes a SSH Agent.
+     *
+     * @throws IOException
+     */
+    private void initRemoteAgent() throws IOException {
+
+        List<SSHUserPrivateKey> userPrivateKeys = new ArrayList<SSHUserPrivateKey>();
+        for (String id : new LinkedHashSet<String>(step.getCredentials())) {
+            final SSHUserPrivateKey c = CredentialsProvider.findCredentialById(id, SSHUserPrivateKey.class, build);
+            if (c == null && !step.isIgnoreMissing()) {
+                listener.fatalError(Messages.SSHAgentBuildWrapper_CredentialsNotFound());
+            }
+            if (c != null && !userPrivateKeys.contains(c)) {
+                userPrivateKeys.add(c);
+            }
+        }
+        for (SSHUserPrivateKey userPrivateKey : userPrivateKeys) {
+            listener.getLogger().println(Messages.SSHAgentBuildWrapper_UsingCredentials(SSHAgentBuildWrapper.description(userPrivateKey)));
+        }
+
+        listener.getLogger().println("[ssh-agent] Looking for ssh-agent implementation...");
+        Map<String, Throwable> faults = new LinkedHashMap<String, Throwable>();
+        for (RemoteAgentFactory factory : Jenkins.getActiveInstance().getExtensionList(RemoteAgentFactory.class)) {
+            if (factory.isSupported(launcher, listener)) {
+                try {
+                    listener.getLogger().println("[ssh-agent]   " + factory.getDisplayName());
+                    agent = factory.start(launcher, listener);
+                    break;
+                } catch (Throwable t) {
+                    faults.put(factory.getDisplayName(), t);
+                }
+            }
+        }
+        if (agent == null) {
+            listener.getLogger().println("[ssh-agent] FATAL: Could not find a suitable ssh-agent provider");
+            listener.getLogger().println("[ssh-agent] Diagnostic report");
+            for (Map.Entry<String, Throwable> fault : faults.entrySet()) {
+                listener.getLogger().println("[ssh-agent] * " + fault.getKey());
+                StringWriter sw = new StringWriter();
+                fault.getValue().printStackTrace(new PrintWriter(sw));
+                for (String line : StringUtils.split(sw.toString(), "\n")) {
+                    listener.getLogger().println("[ssh-agent]     " + line);
```

Comment:
Some of this code looks like it was copied from the `BuildWrapper`. If so, could it be factored out into a common method to avoid duplication?